### PR TITLE
DOC-1565: explain differences in partitioning between PostgreSQL and CRDB

### PR DESCRIPTION
**Pull request by nickvigilante**
Fixes DOC-1565

### CODE REVIEW

**Review by ajwerner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/12706)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nickvigilante and @vy-ton)

___
*[v21.1/partitioning.md, line 713 at r1](https://reviewable.io/reviews/cockroachdb/docs/12706#-Mt63056-Cj7WlmFIDFG:-Mt63056-Cj7WlmFIDFH:boo3zco) ([raw file](https://github.com/cockroachdb/docs/blob/637eddff3c9ad72642ac8c485323b8bf47e4d822/v21.1/partitioning.md#L713)):*
> ```Markdown
> - **Bulk Loading & Deleting:** CockroachDB does not have a feature that supports this use case as of now.
> 
> Whereas a partition in PostgreSQL is effectively its own table, CockroachDB partitions point to the same table but allow the distribution of data across nodes based on geographical and access frequency requirements. CockroachDB allows enabling and disabling partitioning on tables, whereas PostgreSQL currently does not.
> ```

Do these docs not exist in later versions? I feel like we keep parallel sets of docs for the different versions (for better or for worse). 



_Code quote:_
```Markdown
ition in PostgreSQL is effectively its own table, CockroachDB partitions point to the same table but allow the distribution of data across nodes based on geographical and access frequency requirements. CockroachD
```
___
*[v21.1/partitioning.md, line 713 at r1](https://reviewable.io/reviews/cockroachdb/docs/12706#-Mt63QzAD_pMCfkzj9MX:-Mt63QzB4G1amjbWkUiN:b-bbis3i) ([raw file](https://github.com/cockroachdb/docs/blob/637eddff3c9ad72642ac8c485323b8bf47e4d822/v21.1/partitioning.md#L713)):*
> ```Markdown
> - **Bulk Loading & Deleting:** CockroachDB does not have a feature that supports this use case as of now.
> 
> Whereas a partition in PostgreSQL is effectively its own table, CockroachDB partitions point to the same table but allow the distribution of data across nodes based on geographical and access frequency requirements. CockroachDB allows enabling and disabling partitioning on tables, whereas PostgreSQL currently does not.
> ```

I feel like we should also talk about the negative tradeoffs. In postgres, one can leverage partitions for data lifecycle management. Importantly it is cheap to drop partitions, much cheaper than deleting all of the data with `DELETE`. It is also a mechanism to control where the data is stored in terms of postgres's table spaces. You could argue that our partitions can also be used to control the storage medium through the usage of constraints and tags for different stores, but as far as I can tell, nobody really does that. Consider referencing https://github.com/cockroachdb/cockroach/issues/51295 in known limitations. I do think that's an issue we could consider doing something about for 22.2 if it were properly motivated.

**Comment by nickvigilante** [comment]
@ajwerner Yes, there's a 21.2 doc, and I intend to put it in there as well. I just wanted to start the changes in 21.1 so that I can just replicate it to 21.2.

I'll take your feedback on the negative tradeoffs, replicate it to 21.2, and then regenerate the doc for your review.

**Comment by nickvigilante** [comment]
@ajwerner I just updated the docs. Netlify build generating. Can you review once more? If more work is needed on this, can you let me know when you're free for a 1:1 to discuss this a bit further?

@vy-ton FYI

**Comment by nickvigilante** [comment]
@vy-ton How does this look now?

**Review by gemma-shay — APPROVED** [review]
LGTM

**Comment by nickvigilante** [comment]
@gemma-shay TFTR!

**Comment by nickvigilante** [comment]
@ajwerner Can you give one last final review on this?

**Review by lnhsingh — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/12706)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nickvigilante and @vy-ton)

___
*[v21.1/partitioning.md, line 713 at r4](https://reviewable.io/reviews/cockroachdb/docs/12706#-MvTvs554-GV7LIIULCB:-MvTvs554-GV7LIIULCC:b-to07vl) ([raw file](https://github.com/cockroachdb/docs/blob/bd977b58a2202aa5b2b9f2380dc41b7d18f104ba/v21.1/partitioning.md#L713)):*
> ```Markdown
> - **Bulk Loading & Deleting:** CockroachDB does not have a feature that supports this use case as of now.
> 
> PostgreSQL allows its users to leverage partitions for data lifecycle management. Whereas a partition in PostgreSQL is effectively its own table, CockroachDB partitions point to the same table but allow the distribution of data across nodes based on geographical and access frequency requirements. CockroachDB allows enabling and disabling partitioning on tables, whereas PostgreSQL currently does not.
> ```

I'm not sure if we usually say in our docs "This is what Postgres does and this is what CRDB does" since it then requires us to maintain docs on what Postgres (or other specific dbs) do. 

To keep the info in a similar structure to the above bullets, I'd recommend restructuring the info to be something like:

```
- **[I'm not sure what this should be labeled]:** CockroachDB partitions point to the same table but allow the distribution of data across nodes based on geographical and access frequency requirements. This is different from other databases where a partition is effectively its own table. CockroachDB allows enabling and disabling partitioning on tables, whereas other databases may not. Because of this, deleting data within partitions could be expensive resource-wise.
```

^ I'm not confident that I didn't change the technical meaning of what you wrote and would want Vy or Andrew to review.

**Review by ajwerner — COMMENTED** [review]
@devadvocado you should give this a read in light of your recent investigations into this topic.

> Reviewed 3 of 6 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/12706#-:-N801Nil3UecReFn2SWh:bo21vqf)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nickvigilante and @vy-ton)

___
*[`v21.1/partitioning.md` line 712 at r5](https://reviewable.io/reviews/cockroachdb/docs/12706#-N800FdI-CpG7ejsLL1P:-N800FdI-CpG7ejsLL1Q:bu5wj31) ([raw file](https://github.com/cockroachdb/docs/blob/d79d802d8ae9299bc71281f9e64a81e3efd2bfc7/v21.1/partitioning.md#L712)):*
> Because of this, deleting data within partitions could be expensive resource-wise.

I think I understand what you mean to say, but I don't think this sentence is going to help aid in understanding. Maybe say:

```
CockroachDB uses the partitioning concept to allow users to logically subdivide a single tables physical table to enable independent configuration of those partitions using zone configurations. Other databases sometimes implement partitioning as the logical union of physically separate tables. This difference means that cockroach is able to permit inexpensive enabling, disabling, and modification of partition boundaries in contrast to other databases. Consequently cockroach is unable to provide the same bulk data deletion operations over table partitions which the other databases achieve by physically dropping the underlying table represented by the partition. 
```



_Code quote:_
```Markdown
 **Logical structure of partitions:** CockroachDB partitions point to the same table but allow the distribution of data across nodes based on geographical and access frequency requirements. This is different from other databases where a partition is effectively its own table. CockroachDB allows enabling and disabling partitioning on tables, whereas other databases may not. Because of this, deleting data within partitions could be expensive resource-wise.
```

**Review by nickvigilante — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/12706)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @lnhsingh, and @vy-ton)

___
*[`v21.1/partitioning.md` line 713 at r1](https://reviewable.io/reviews/cockroachdb/docs/12706#-Mt63056-Cj7WlmFIDFG:-N807eka80AG_y3ScOAx:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/637eddff3c9ad72642ac8c485323b8bf47e4d822/v21.1/partitioning.md#L713)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Do these docs not exist in later versions? I feel like we keep parallel sets of docs for the different versions (for better or for worse). 
</blockquote></details>

Done.
___
*[`v21.1/partitioning.md` line 713 at r1](https://reviewable.io/reviews/cockroachdb/docs/12706#-Mt63QzAD_pMCfkzj9MX:-N807fGe-AYj3nKgOsXP:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/637eddff3c9ad72642ac8c485323b8bf47e4d822/v21.1/partitioning.md#L713)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I feel like we should also talk about the negative tradeoffs. In postgres, one can leverage partitions for data lifecycle management. Importantly it is cheap to drop partitions, much cheaper than deleting all of the data with `DELETE`. It is also a mechanism to control where the data is stored in terms of postgres's table spaces. You could argue that our partitions can also be used to control the storage medium through the usage of constraints and tags for different stores, but as far as I can tell, nobody really does that. Consider referencing https://github.com/cockroachdb/cockroach/issues/51295 in known limitations. I do think that's an issue we could consider doing something about for 22.2 if it were properly motivated. 
</blockquote></details>

Done.
___
*[`v21.1/partitioning.md` line 713 at r4](https://reviewable.io/reviews/cockroachdb/docs/12706#-MvTvs554-GV7LIIULCB:-N807fc24MVZ3a-ob55A:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/bd977b58a2202aa5b2b9f2380dc41b7d18f104ba/v21.1/partitioning.md#L713)):*
<details><summary><i>Previously, lnhsingh (Lauren Hirata Singh) wrote…</i></summary><blockquote>

I'm not sure if we usually say in our docs "This is what Postgres does and this is what CRDB does" since it then requires us to maintain docs on what Postgres (or other specific dbs) do. 

To keep the info in a similar structure to the above bullets, I'd recommend restructuring the info to be something like:

```
- **[I'm not sure what this should be labeled]:** CockroachDB partitions point to the same table but allow the distribution of data across nodes based on geographical and access frequency requirements. This is different from other databases where a partition is effectively its own table. CockroachDB allows enabling and disabling partitioning on tables, whereas other databases may not. Because of this, deleting data within partitions could be expensive resource-wise.
```

^ I'm not confident that I didn't change the technical meaning of what you wrote and would want Vy or Andrew to review.

</blockquote></details>

Done.
___
*[`v21.1/partitioning.md` line 712 at r5](https://reviewable.io/reviews/cockroachdb/docs/12706#-N800FdI-CpG7ejsLL1P:-N807gTd15caEwxhY9tK:b-bw8e14) ([raw file](https://github.com/cockroachdb/docs/blob/d79d802d8ae9299bc71281f9e64a81e3efd2bfc7/v21.1/partitioning.md#L712)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> Because of this, deleting data within partitions could be expensive resource-wise.

I think I understand what you mean to say, but I don't think this sentence is going to help aid in understanding. Maybe say:

```
CockroachDB uses the partitioning concept to allow users to logically subdivide a single tables physical table to enable independent configuration of those partitions using zone configurations. Other databases sometimes implement partitioning as the logical union of physically separate tables. This difference means that cockroach is able to permit inexpensive enabling, disabling, and modification of partition boundaries in contrast to other databases. Consequently cockroach is unable to provide the same bulk data deletion operations over table partitions which the other databases achieve by physically dropping the underlying table represented by the partition. 
```
</blockquote></details>

Updated.

**Comment by nickvigilante** [comment]
@devadvocado Can you review?

**Review by devadvocado — APPROVED** [review]
:lgtm:

> Reviewed 2 of 6 files at r5, 3 of 3 files at r6, 3 of 3 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/12706#-:-NCVv14Q2x8FnOfmvSb3:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner and @lnhsingh)

File: v22.2/partitioning.md [hunk 1]
```diff
@@ -716,11 +716,14 @@ Other databases use partitioning for three additional use cases: secondary index
...
 - **Sharding:** CockroachDB automatically shards data as a part of its distributed database architecture.
 - **Bulk Loading & Deleting:** CockroachDB does not have a feature that supports this use case as of now.
+- **Logical structure of partitions:** CockroachDB uses the partitioning concept to allow users to logically subdivide a single physical table to enable independent configuration of those partitions using zone configurations. Other databases sometimes implement partitioning as the logical union of physically separate tables. This difference means that CockroachDB is able to permit inexpensive repartitioning in contrast to other databases. Consequently, CockroachDB is unable to provide the same bulk data deletion operations over table partitions that other databases achieve by physically dropping the underlying table represented by the partition.
 
 ## Known limitations
 
 - {% include {{ page.version.version }}/known-limitations/partitioning-with-placeholders.md %}
 
+- CockroachDB does not currently support dropping a single partition from a table. In order to remove partitions, you can [repartition](#repartition-a-table) the table.
```
Comment (lnhsingh) [inline v22.2/partitioning.md:725]: This known limitation should be in an include and also added to to the corresponding `<version>/known-limitations.md` doc (this applies across all versions)

File: v21.2/partitioning.md [hunk 2]
```diff
@@ -22,10 +21,10 @@ Table partitioning helps you reduce latency and cost:
 
 Table partitioning involves a combination of CockroachDB features:
 
-- [Node attributes](#node-attributes)
-- [Enterprise license](#enterprise-license)
-- [Table creation](#table-creation)
-- [Replication zones](#replication-zones)
+- [Node Attributes](#node-attributes)
```
Comment (lnhsingh) [inline v21.2/partitioning.md:24]: All of these bullets should be in sentence case

File: v21.2/partitioning.md [hunk 3]
```diff
@@ -99,13 +98,13 @@ For instance, consider the database of a global online learning portal that has
     PRIMARY KEY (country, id));
 ~~~
 
-##### Primary key considerations
+**Primary Key Considerations**
```
Comment (lnhsingh) [inline v21.2/partitioning.md:101]: This should still be sentence case

File: v21.2/partitioning.md [hunk 4]
```diff
@@ -543,7 +536,7 @@ You should see the following output:
 (2 rows)
 ~~~
 
-### Define subpartitions on a table
+### Define subpartitions on a Table
```
Comment (lnhsingh) [inline v21.2/partitioning.md:539]: All of these H3s should be in sentence case

**Review by lnhsingh — COMMENTED** [review]
Comments apply across versions

**Review by nickvigilante — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/12706)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner, @devadvocado, and @lnhsingh)

___
*[`v21.2/partitioning.md` line 24 at r7](https://reviewable.io/reviews/cockroachdb/docs/12706#-MtA28frtghKtRhr5hPV-r7-24:-NCVzz90BbdnMJmLe3TU:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/805df05a543506ac5b73d85ffcf7740db7c21b0a/v21.2/partitioning.md#L24)):*
<details><summary><i>Previously, lnhsingh (Lauren Hirata Singh) wrote…</i></summary><blockquote>

All of these bullets should be in sentence case
</blockquote></details>

Done.
___
*[`v21.2/partitioning.md` line 101 at r7](https://reviewable.io/reviews/cockroachdb/docs/12706#-MtA28frtghKtRhr5hPV-r7-101:-NCW-AdYCqb-liItzLPD:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/805df05a543506ac5b73d85ffcf7740db7c21b0a/v21.2/partitioning.md#L101)):*
<details><summary><i>Previously, lnhsingh (Lauren Hirata Singh) wrote…</i></summary><blockquote>

This should still be sentence case
</blockquote></details>

Done.



_Code quote:_
```Markdown
##### Primary key considerations
```
___
*[`v21.2/partitioning.md` line 539 at r7](https://reviewable.io/reviews/cockroachdb/docs/12706#-MtA28frtghKtRhr5hPV-r7-539:-NCW-LlI93BKrcwMpjEi:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/805df05a543506ac5b73d85ffcf7740db7c21b0a/v21.2/partitioning.md#L539)):*
<details><summary><i>Previously, lnhsingh (Lauren Hirata Singh) wrote…</i></summary><blockquote>

All of these H3s should be in sentence case
</blockquote></details>

Done.
___
*[`v22.2/partitioning.md` line 725 at r7](https://reviewable.io/reviews/cockroachdb/docs/12706#-NCVU0xAr_ag5QM9cdaU-r7-725:-NCW5pKq-Vbl_XD45wPc:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/805df05a543506ac5b73d85ffcf7740db7c21b0a/v22.2/partitioning.md#L725)):*
<details><summary><i>Previously, lnhsingh (Lauren Hirata Singh) wrote…</i></summary><blockquote>

This known limitation should be in an include and also added to to the corresponding `<version>/known-limitations.md` doc (this applies across all versions)
</blockquote></details>

Done.

**Review by lnhsingh — APPROVED** [review]

**Pull request merged**